### PR TITLE
[MAINTENANCE] Add typevar around add_expectation

### PR DESCRIPTION
**Pull request by tyler-hoffman**
Adds TypeVar so that the type checker knows we return the same subclass of Expectation that we pass in to add_expectation.

- [ ] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [ ] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB]
- [ ] Code is linted - run `invoke lint` (uses `black` + `ruff`)
- [ ] Appropriate tests and docs have been updated

For more information about contributing, see [Contribute](https://docs.greatexpectations.io/docs/contributing/contributing_checklist).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

### CODE REVIEW

File: great_expectations/data_context/store/expectations_store.py [hunk 1]
```diff
@@ -28,6 +28,8 @@
 if TYPE_CHECKING:
     from great_expectations.expectations.expectation import Expectation
 
+    _TExpectation = TypeVar("_TExpectation", bound=Expectation)
```
Comment (joshua-stauffer) [inline great_expectations/data_context/store/expectations_store.py:31]: nice!

File: great_expectations/core/expectation_suite.py [hunk 2]
```diff
@@ -562,7 +565,8 @@ def find_expectation_indexes(
                     match_indexes.append(idx)
             else:  # noqa: PLR5501
                 if expectation.configuration.isEquivalentTo(
-                    other=expectation_configuration, match_type=match_type  # type: ignore[arg-type]
+                    other=expectation_configuration,
+                    match_type=match_type,  # type: ignore[arg-type]
                 ):
```
Comment (joshua-stauffer) [inline great_expectations/core/expectation_suite.py]: i'm confused - why did this change?

Comment (joshua-stauffer) [reply great_expectations/core/expectation_suite.py]: ah, looks like you can just remove this type ignore

**Review by joshua-stauffer — APPROVED** [review]
looks great, thanks!

File: great_expectations/core/expectation_suite.py (hunk 2 above)
Comment (tyler-hoffman) [reply great_expectations/core/expectation_suite.py]: yeah, looks like autoformat moved it and it can go 🎉

**Review by billdirks — APPROVED** [review]
LGTM!

**Review by Rachel-Reverie — APPROVED** [review]
LGTM

**Pull request merged**